Add CancellationToken overloads for reading multipart form sections

martincostello opened the pull request:
# Add CancellationToken overloads for reading multipart form sections

Support passing through a CancellationToken when reading multipart forms.

## Description

- Add `CancellationToken` overloads to methods for reading multipart form sections and use in `FormFeature`.
- Update `MultipartSectionStreamExtensions` to apply IDE refactoring suggestions.

Fixes #41532.

martincostello committed:
Add CancellationToken overloads

Add CancellationToken overloads to methods for reading multipart form sections and use in FormFeature.
Update MultipartSectionStreamExtensions to apply IDE refactoring suggestions.
Relates to #41532.

campersau commented on src/Http/WebUtilities/src/MultipartSectionStreamExtensions.cs:
```diff
@@ -17,18 +17,24 @@ public static class MultipartSectionStreamExtensions
     /// <param name="section">The section to read from</param>
     /// <returns>The body steam as string</returns>
     public static async Task<string> ReadAsStringAsync(this MultipartSection section)
+        => await section.ReadAsStringAsync(CancellationToken.None);
```
```suggestion
    public static Task<string> ReadAsStringAsync(this MultipartSection section)
        => section.ReadAsStringAsync(CancellationToken.None);
```

martincostello replied:
I preserved the await because the method it calls also doesn't just return the raw `Task`.

martincostello replied:
With the refactor to remove the block, shall I remove both then if/when the API changes are approved?

Tratcher replied:
The other method still needs the await so the using statement doesn't dispose the reader too soon. Please remove the await from this overload though.

Tratcher commented on src/Http/WebUtilities/src/FormMultipartSection.cs:
```diff
@@ -53,8 +53,15 @@ public FormMultipartSection(MultipartSection section, ContentDispositionHeaderVa
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The form value</returns>
+    public Task<string> GetValueAsync(CancellationToken cancellationToken)
```
```suggestion
    public ValueTask<string> GetValueAsync(CancellationToken cancellationToken)
```

Tratcher commented on src/Http/WebUtilities/src/MultipartSectionStreamExtensions.cs:
```diff
@@ -17,18 +17,24 @@ public static class MultipartSectionStreamExtensions
+    /// <param name="section">The section to read from</param>
+    /// <param name="cancellationToken">The cancellationt token.</param>
+    /// <returns>The body steam as string</returns>
+    public static async Task<string> ReadAsStringAsync(this MultipartSection section, CancellationToken cancellationToken)
```
```suggestion
    public static async ValueTask<string> ReadAsStringAsync(this MultipartSection section, CancellationToken cancellationToken)
```

martincostello committed:
API review feedback

- Change return types to ValueTask<string>.
- Use expression bodied method.

martincostello committed:
Use ArgumentNullException.ThrowIfNull

- Use ArgumentNullException.ThrowIfNull()
- Resolve warning about unused return value.
